### PR TITLE
Make convenient functions `is_nan`, `is_inf`

### PR DESCRIPTION
**Pull request by fehiepsi**
Besides the above convenient functions, I also clean HMC a bit.

### CODE REVIEW

File: pyro/util.py [hunk 1]
```diff
@@ -181,6 +182,26 @@ def ng_zeros(*args, **kwargs):
...
 
 
+def is_nan(x):
+    """
+    A convenient function to check if a Tensor contains all nan; also works with numbers
+    and torch.autograd.Variable
+    """
+    if isinstance(x, numbers.Number):
+        return x != x
+    return all(x != x)
```
Comment (fritzo) [inline pyro/util.py]: I believe this should use PyTorch via
```py
return (x !=  x).all()
```

File: pyro/util.py [hunk 2]
```diff
@@ -181,6 +182,26 @@ def ng_zeros(*args, **kwargs):
...
+
+
+def is_inf(x):
+    """
+    A convenient function to check if a Tensor contains all inf; also works with numbers
+    and torch.autograd.Variable
+    """
+    if isinstance(x, numbers.Number):
+        return x == float('inf')
+    return all(x == float('inf'))
```
Comment (fritzo) [inline pyro/util.py]: ditto:
```py
return (x == float('inf')).all()
```

File: pyro/util.py (hunk 1 above)
Comment (fehiepsi) [reply pyro/util.py]: @fritzo It seems that they are the same in my system. Anyway, I made the changes based on my your suggestions.

Comment (fritzo) [reply pyro/util.py]:
```py
>>> all(torch.ones(2))
True
>>> all(torch.ones(2, 2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/fritzobermeyer/pytorch-dev/2.7/torch/tensor.py", line 168, in __bool__
    " containing more than one value is ambiguous")
RuntimeError: bool value of torch.FloatTensor containing more than one value is ambiguous
```

**Pull request merged**